server/src/main/java/io/druid/segment/realtime/firehose/ReplayableFirehoseFactory.java

```diff
@@ -0,0 +1,313 @@
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Creates a wrapper firehose that writes from another firehose to disk and then serves nextRow() from disk. Useful for
+ * tasks that require multiple passes through the data to prevent multiple remote fetches. Also has support for
+ * retrying fetches if the underlying firehose throws an exception while the local cache is being generated.
+ */
+public class ReplayableFirehoseFactory implements FirehoseFactory<InputRowParser>, Closeable
+{
+  private static final EmittingLogger log = new EmittingLogger(ReplayableFirehoseFactory.class);
+  private static final boolean DEFAULT_REPORT_PARSE_EXCEPTIONS = false;
+  private static final int DEFAULT_MAX_TEMP_FILE_SIZE = 250000000;
+  private static final int DEFAULT_READ_FIREHOSE_RETRIES = 3;
+
+  private final FirehoseFactory delegateFactory;
+  private final boolean reportParseExceptions;
+  private final int maxTempFileSize;
+  private final int readFirehoseRetries;
+
+  private ReplayableFirehose firehose;
+
+  @JsonCreator
+  public ReplayableFirehoseFactory(
+      @JsonProperty("delegate") FirehoseFactory delegateFactory,
+      @JsonProperty("reportParseExceptions") Boolean reportParseExceptions,
+      @JsonProperty("maxTempFileSize") Integer maxTempFileSize,
+      @JsonProperty("readFirehoseRetries") Integer readFirehoseRetries
+  )
+  {
+    Preconditions.checkNotNull(delegateFactory, "delegate cannot be null");
+    Preconditions.checkArgument(
+        !(delegateFactory instanceof ReplayableFirehoseFactory),
+        "Refusing to wrap another ReplayableFirehoseFactory"
+    );
+
+    this.delegateFactory = delegateFactory;
+    this.reportParseExceptions = reportParseExceptions == null
+                                 ? DEFAULT_REPORT_PARSE_EXCEPTIONS
+                                 : reportParseExceptions;
+    this.maxTempFileSize = maxTempFileSize == null ? DEFAULT_MAX_TEMP_FILE_SIZE : maxTempFileSize;
+    this.readFirehoseRetries = readFirehoseRetries == null ? DEFAULT_READ_FIREHOSE_RETRIES : readFirehoseRetries;
+
+    log.info(this.toString());
+  }
+
+  @Override
+  public Firehose connect(InputRowParser parser) throws IOException
+  {
+    if (firehose == null) {
+      firehose = new ReplayableFirehose(parser);
+    } else {
+      log.info("Rewinding and returning existing firehose");
+      firehose.rewind();
+    }
+
+    return firehose;
+  }
+
+  @Override
+  public void close() throws IOException
+  {
+    if (firehose != null) {
+      firehose.deleteTempFiles();
+      firehose.close();
+    }
+
+    firehose = null;
+  }
+
+  public class ReplayableFirehose implements Firehose
+  {
+    private final JsonFactory factory = new DefaultObjectMapper(
+        new SmileFactory().enable(SmileGenerator.Feature.CHECK_SHARED_STRING_VALUES)
+    ).getFactory();
```

Comment:
Oops, no this doesn't work with custom serializers; I have to use the Guice managed one through `@Smile @JacksonInject`; nice catch, thank you.